include/nudb/basic_store.hpp

```diff
@@ -343,6 +361,61 @@ class basic_store
         error_code& ec,
         Args&&... args);
 
+    /** Open a database.
+
+        The database identified by the specified directory
+        under which default data and key paths are opened.
+
+        @par Requirements
+
+        The database must be not be open.
+
+        @par Thread safety
+
+        Not thread safe. The caller is responsible for
+        ensuring that no other member functions are
+        called concurrently.
+
+        @param dir_path The path to the directory containing data
+        and key files.
+
+        @param ec Set to the error, if any occurred.
+
+        @param args Optional arguments passed to @b File constructors.
+
+    */
+    template<class... Args>
+    void
+    open(
+        path_type const& dir_path,
+        error_code& ec,
+        Args&&... args);
+
+    /** Open a database in read only mode.
+
+         Takes same parameters as open but opens
+         all files in read mode and disallows
+         insertions.
+
+    */
+    template<class... Args>
+    void
+    open_read_only(
+        path_type const& dat_path,
+        path_type const& key_path,
+        error_code& ec,
+        Args&&... args);
+
+    /** Open a database directory in read only mode.
+
+    */
+    template<class... Args>
+    void
+    open_read_only(
```

Comment:
To avoid a combinatorial explosion of signatures, the open mode should be a parameter.